packages/aws-cdk-lib/aws-synthetics/lib/canary.ts

```diff
@@ -669,37 +669,44 @@ export class Canary extends cdk.Resource implements ec2.IConnectable {
   }
 
   private createArtifactConfig(props: CanaryProps): CfnCanary.ArtifactConfigProperty | undefined {
-    if (!props.artifactS3EncryptionMode) {
+    if (!props.artifactS3EncryptionMode && !props.artifactS3KmsKey) {
       return undefined;
     }
 
     const isNodeRuntime = props.runtime.family === RuntimeFamily.NODEJS;
-    const isArtifactS3EncryptionModeDefined = !cdk.Token.isUnresolved(props.artifactS3EncryptionMode) && props.artifactS3EncryptionMode;
-    const isArtifactS3KmsKeyDefined = !cdk.Token.isUnresolved(props.artifactS3KmsKey) && props.artifactS3KmsKey;
 
     if (
-      isArtifactS3EncryptionModeDefined &&
-      props.artifactS3EncryptionMode !== ArtifactsEncryptionMode.KMS &&
-      isArtifactS3KmsKeyDefined
+      props.artifactS3EncryptionMode === ArtifactsEncryptionMode.S3_MANAGED &&
+      props.artifactS3KmsKey
     ) {
-      throw new Error('A customer-managed KMS key was provided, but the encryption mode is not set to SSE-KMS.');
+      throw new Error(`A customer-managed KMS key was provided, but the encryption mode is not set to SSE-KMS, got: ${props.artifactS3EncryptionMode}.`);
     }
 
     // Only check runtime family is Node.js because versions prior to `syn-nodejs-puppeteer-3.3` are deprecated and can no longer be configured.
-    if (!isNodeRuntime && isArtifactS3EncryptionModeDefined) {
+    if (!isNodeRuntime && props.artifactS3EncryptionMode) {
       throw new Error(`Artifact encryption is only supported for canaries that use Synthetics runtime version \`syn-nodejs-puppeteer-3.3\` or later, got \`${props.runtime.name}\`.`);
     }
 
     let encryptionKey: kms.IKey | undefined;
-    if (props.artifactS3EncryptionMode === ArtifactsEncryptionMode.KMS) {
-      encryptionKey = props.artifactS3KmsKey ?? new kms.Key(this, 'Key', { description: `Created by ${this.node.path}` });
+    if (props.artifactS3EncryptionMode === ArtifactsEncryptionMode.KMS && !props.artifactS3KmsKey) {
+      encryptionKey = new kms.Key(this, 'Key', { description: `Created by ${this.node.path}` });
+    } else {
+      encryptionKey = props.artifactS3KmsKey;
     }
```

Comment:
I'm curious about why we made this change. The previous lines seemed more intuitive in my opinion, whereas the second isn't quite as clear (e.g. `encryptionKey` will still be set if encryption mode is S3, since it'll fail the first part of the condition). If we need the default behaviour to be that the encryption mode is set to KMS when a key is provided, maybe we could instead set the encryption mode before this block? (i.e. move lines L699-L706 before this)

Reply:
Thank you. I reviewed it to set the `encryptionMode` first, followed by setting the `encryptionKey`.